# Random Forest Classifier carbon, silicon, and nitrogen isotopes
TEST NOTEBOOK, trying to apply cross tabs code from:
http://www.agcross.com/2015/02/random-forests-scikit-learn/
need Categorical.from_codes code:
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Categorical.from_codes.html


In [3]:
#Imports
import joblib
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

from sklearn import tree
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, MinMaxScaler


In [18]:
# Read csv file
C_Si_N = pd.read_csv('presolargrains_C_Si_N.csv')
C_Si_N.head()

Type  carbon_12_13  nitrogen_14_15  silicon_29_28  silicon_30_28
0    X         74.30           207.4         -685.0         -520.0
1    X       1581.00           116.0         -684.0         -490.0
2    X        234.75           187.2         -683.0         -501.0
3    X        455.00           140.0         -662.0         -770.0
4    X        823.00            44.0         -658.0         -234.0

In [6]:
C_Si_N['is_train'] = np.random.uniform(0, 1, len(C_Si_N)) <= .8

In [19]:
'''
# Label Encoder and Transform
label_encoder = LabelEncoder()

data = C_Si_N['Type'].copy()

label_encoder.fit(data)
encoded_y = label_encoder.transform(data)
label_encoder.classes_
encoded_y

target = C_Si_N['Type']
target_names = ['1', '2', '3', '4', '5', '6', '7', '8']
print(target_names)
'''

"\n# Label Encoder and Transform\nlabel_encoder = LabelEncoder()\n\ndata = C_Si_N['Type'].copy()\n\nlabel_encoder.fit(data)\nencoded_y = label_encoder.transform(data)\nlabel_encoder.classes_\nencoded_y\n"

In [21]:
target = C_Si_N['Type']
print('target:', target)
target_names = ['AB', 'C', 'M', 'N', 'U', 'X', 'Y', 'Z']
print('target names', target_names)

target: 0        X
1        X
2        X
3        X
4        X
5        X
6        X
7        X
8        X
9        X
10       X
11       X
12       X
13       X
14       X
15       X
16       X
17       X
18       X
19       X
20       X
21       X
22       X
23       N
24       X
25       X
26       X
27       X
28       X
29       X
        ..
1271     M
1272     M
1273     M
1274     M
1275     M
1276     M
1277     M
1278     M
1279     M
1280     M
1281    AB
1282     M
1283     M
1284     M
1285     M
1286    AB
1287     M
1288    AB
1289     X
1290    AB
1291     M
1292     U
1293     M
1294     M
1295     C
1296     C
1297     C
1298     C
1299     C
1300     C
Name: Type, Length: 1301, dtype: object
target names ['AB', 'C', 'M', 'N', 'U', 'X', 'Y', 'Z']


In [29]:
print(len(target_names))

8


In [26]:
C_Si_N['species'] = pd.Categorical.from_codes(codes=[0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1], categories=target_names)

ValueError: Length of values does not match length of index

In [11]:
# Set up data and feature names; display dataframe dropping columns
data = C_Si_N.drop("Type", axis=1)
feature_names = data.columns
data.head()

carbon_12_13  nitrogen_14_15  silicon_29_28  silicon_30_28
0         74.30           207.4         -685.0         -520.0
1       1581.00           116.0         -684.0         -490.0
2        234.75           187.2         -683.0         -501.0
3        455.00           140.0         -662.0         -770.0
4        823.00            44.0         -658.0         -234.0

In [12]:
target = C_Si_N['Type']

In [13]:
# Separate data into train and test buckets
X_train, X_test, y_train, y_test = train_test_split(data, target, random_state=1)

## Scale the data using MinMaxScalar

In [14]:
X_scaler = MinMaxScaler().fit(X_train)
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Train the model; run the test score

In [15]:
# Random Forest Classifier
rf = RandomForestClassifier(n_estimators=200)
rf = rf.fit(X_train_scaled, y_train)
rf.score(X_test_scaled, y_test)

0.9693251533742331

In [16]:
# Save fitted model to file
filename = 'RandomForestCls_C_Si_N.sav'
joblib.dump(rf, filename)

['RandomForestCls_C_Si_N.sav']

## Summary:

### Random Forest Classifier carbon, silicon, and nitrogen:  96.9%